### ISSUE TITLE
Not reading “label” field when initializing SFTTrainer

### ISSUE THREAD

**Issue by paihengxu** [huggingface/trl#689]
When I use SFFTrainer to fine-tune a LM for sequence classification, the SFTTrainer does not read the "label" field in the dataset I passed.

The dataset I used was in the type of `datasets.arrow_dataset.Dataset` from the `datasets` package. I was able to convert it to torch dataset using fix from [here](https://github.com/huggingface/datasets/issues/2165), so it does nothing and directly uses the dataset I passed according to  https://github.com/huggingface/trl/blob/d3bbee3ab84d4d5dcb413003bf93541df5b4e54c/trl/trainer/sft_trainer.py#L249-L251

I am just wondering if it's a feature or bug to (1) not support `datasets.arrow_dataset.Dataset` and (2) not loading the "label" field in https://github.com/huggingface/trl/blob/d3bbee3ab84d4d5dcb413003bf93541df5b4e54c/trl/trainer/sft_trainer.py#L305 and https://github.com/huggingface/trl/blob/d3bbee3ab84d4d5dcb413003bf93541df5b4e54c/trl/trainer/utils.py#L506

**Comment by younesbelkada** [comment]
Hi @paihengxu 
Thanks for the issue, can you also provide a handy snippet that best describes the issue?

**Comment by paihengxu** [comment]
Hi @younesbelkada, sure, here it is.

```from datasets import load_dataset
from transformers import (
    AutoModelForSequenceClassification,
    AutoTokenizer,
    TrainingArguments
)
from trl import SFTTrainer


model_name = "roberta-large"
dataset = load_dataset('imdb', split="train")

# Load base model
model = AutoModelForSequenceClassification.from_pretrained(
    model_name,
    num_labels=2
)
tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)


# Set training parameters
training_arguments = TrainingArguments(
    output_dir="test",
    num_train_epochs=1,
    per_device_train_batch_size=4,
    save_strategy="epoch",
    logging_strategy="epoch",
    learning_rate=2e-5,
    weight_decay=0.01,
    report_to="none",
    push_to_hub=False,
)

# Set supervised fine-tuning parameters
trainer = SFTTrainer(
    model=model,
    train_dataset=dataset,
    dataset_text_field="text",
    tokenizer=tokenizer,
    args=training_arguments,
    packing=False,
)

# Train model
trainer.train()
```

**Comment by wanxingliu94** [comment]
encountered the same issue, any update on this?

**Comment by paihengxu** [comment]
> encountered the same issue, any update on this?

I was able to get it working by converting the dataset to a Torch dataset shown in the fix I linked above.

**Comment by lvwerra** [comment]
The SFTTrainer only works with the language modeling objective, not with sequence classification. Use the `transformers.Trainer` for this.

**Comment by sujantkumarkv** [comment]
hi,

my doubt is regarding the `dataset_text_field="text"` parameter in `SFTTrainer`

like the field has to be `text` but I have a different dataset (jsonl file), [here](https://huggingface.co/datasets/sujantkumarkv/indian_legal_corpus); its basically "sentence completion" with "input" & "target" completing the text sequence.

I'm not sure how to make it work with the `dataset_text_field`.

Any help/comment is truly appreciated. 
thanks :D

**Comment by lvwerra** [comment]
In that case you would just need to merge the two columns. For example with a map operation:
```python
ds = ds.map(lambda x: {"text": x["input"]+x["output"]})
```
Hope that helps!

**Comment by George-du** [comment]
Dear @paihengxu,@lvwerra 

        Can this function be used normally for text classification at this moment?

**Comment by younesbelkada** [comment]
Hello @George-du 
SFTTrainer cannot be used for text classification, the trainer can be only used for supervised fine-tuning on text generation (next token prediction)

**Comment by sujantkumarkv** [comment]
> In that case you would just need to merge the two columns. For example with a map operation:
> 
> ```python
> ds = ds.map(lambda x: {"text": x["input"]+x["output"]})
> ```
> 
> Hope that helps!

thanks ❤️

**Comment by George-du** [comment]
Dear  @younesbelkada @lvwerra @paihengxu ,

    Thansk, Is it possible to change SFTTrainer so that it can be used to fine-tune GPT2ForSequenceClassification?

**Comment by younesbelkada** [comment]
Hi @George-du , currently this is not supported, for that one may need to create a new trainer `SeqClsTrainer` as it is out of scope of Supervised Fine-tuning

**Comment by George-du** [comment]
wow! I'm looking forward to this.



At 2023-09-18 19:17:19, "Younes Belkada" ***@***.***> wrote:

Hi @George-du , currently this is not supported, for that one may need to create a new trainer SeqClsTrainer as it is out of scope of Supervised Fine-tuning

**Comment by lvwerra** [comment]
Closing this for now as the original issue seems to be fixed.

**Comment by ashioyajotham** [comment]
> The SFTTrainer only works with the language modeling objective, not with sequence classification. Use the `transformers.Trainer` for this.

Thank you. Worked for me.

**Comment by alexcoca** [comment]
> Hi @George-du , currently this is not supported, for that one may need to create a new trainer `SeqClsTrainer` as it is out of scope of Supervised Fine-tuning

@younesbelkada Does the SFTTrainer support encoder-decoder architectures or only decoder only? Quick glance through the source made me think it only works for decoder-only models.
